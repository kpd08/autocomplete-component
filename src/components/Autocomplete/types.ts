enum Keys {
  ArrowDown = "ArrowDown",
  ArrowUp = "ArrowUp",
  Enter = "Enter",
}

enum EntryType {
  Repo = "repo",
  User = "user",
}

interface GithubUser {
  login: string;
  html_url: string;
}

interface GithubRepo {
  name: string;
  html_url: string;
}

interface AutcompleteEntry {
  name: string;
  type: EntryType;
  url: string;
}

interface GithubSearchResponse<T = any> {
  total_count: number;
  items: Array<T>;
  incomplete_results: boolean;
}

export type { AutcompleteEntry, GithubUser, GithubSearchResponse, GithubRepo };
export { Keys, EntryType };
