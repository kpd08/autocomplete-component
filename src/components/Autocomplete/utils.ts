import { MAX_ENTRIES_COUNT } from "./settings";
import { GithubRepo, GithubUser, AutcompleteEntry, EntryType } from "./types";

const mapUsersAndReposToResults = (
  users: Array<GithubUser> = [],
  repos: Array<GithubRepo> = []
): Array<AutcompleteEntry> => {
  const usersResults: Array<AutcompleteEntry> = users.map((user) => ({
    name: user.login,
    type: EntryType.User,
    url: user.html_url,
  }));

  const reposResults: Array<AutcompleteEntry> = repos.map((repo) => ({
    name: repo.name,
    type: EntryType.Repo,
    url: repo.html_url,
  }));

  return usersResults
    .concat(reposResults)
    .sort((a, b) => a.name.localeCompare(b.name))
    .slice(0, MAX_ENTRIES_COUNT);
};

export { mapUsersAndReposToResults };
