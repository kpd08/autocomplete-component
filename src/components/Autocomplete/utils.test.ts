import { GithubRepo, GithubUser } from "./types";
import { mapUsersAndReposToResults } from "./utils";

const mockRepos: Array<GithubRepo> = [
  { name: "a_repo", html_url: "mock.url" },
  { name: "b_repo", html_url: "mock.url" },
];

const mockUsers: Array<GithubUser> = [
  { login: "a_user", html_url: "mock.url" },
  { login: "b_user", html_url: "mock.url" },
];

describe("mapUsersToResponse", () => {
  it("should map users and repos to result", () => {
    const results = mapUsersAndReposToResults(mockUsers, mockRepos);
    expect(results).toMatchSnapshot();
  });

  it("should return empty array if args are not provided", () => {
    const results = mapUsersAndReposToResults();
    expect(results).toEqual([]);
  });

  it("should limit result to 50 entries", () => {
    const userLong = new Array(100).fill(null).map(() => mockUsers[0]);
    expect(mapUsersAndReposToResults(userLong, mockRepos)).toHaveLength(50);
  });
});
