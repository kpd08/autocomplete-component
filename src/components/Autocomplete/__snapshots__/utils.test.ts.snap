// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapUsersToResponse should map users and repos to result 1`] = `
Array [
  Object {
    "name": "a_repo",
    "type": "repo",
    "url": "mock.url",
  },
  Object {
    "name": "a_user",
    "type": "user",
    "url": "mock.url",
  },
  Object {
    "name": "b_repo",
    "type": "repo",
    "url": "mock.url",
  },
  Object {
    "name": "b_user",
    "type": "user",
    "url": "mock.url",
  },
]
`;
